### PR TITLE
BUGFIX: TNT-1665 New width selected columns

### PR DESCRIPTION
**Pull request by NestorEncinas**
Update `newWidhtForSelectedColumns` to accept an array of measure so columnmanual resizing works properly on new table transposition.

Also add missing `newTotalColumnLocator` to pivot table factories.

JIRA: TNT-1665

---

Supported PR commands:

| Command                                                 | Description                                                |
| ------------------------------------------------------- | ---------------------------------------------------------- |
| `ok to test`                                            | Re-run standard checks                                     |
| `extended check sonar`                                  | SonarQube tests                                            |
| `extended test - backstop`                              | BackstopJS tests                                           |
| **E2E Cypress tests commands - TIGER**                  |                                                            |
| `extended test - tiger-cypress - isolated <testName>`   | Run isolated tests running against recorded Tiger backend. |
| `extended test - tiger-cypress - record <testName>`     | Create a new recording for isolated Tiger tests.           |
| `extended test - tiger-cypress - integrated <testName>` | Run integrated tests against live backend                  |
| **E2E Cypress tests commands - BEAR**                   |                                                            |
| `extended test - cypress - isolated <testName>`         | Run isolated tests running against recorded Bear backend.  |
| `extended test - cypress - record <testName>`           | Create a new recording for isolated Bear tests.            |
| `extended test - cypress - integrated <testName>`       | Run integrated tests against live backend                  |
| **Compatibility matrix test commands - TIGER Backend**  |                                                            |
| `extended test - matrix-test <AIO_version>`             | Run integrated tests against AIO versions.                 |

`<testName>` in cypress commands is used to filter specfiles. Example, to run record with BEAR backend

-   Against `dashboard.spec.ts` and `drilling.spec.ts`, execute command `extended test - cypress - record dashboard,drilling`
-   Against all specfiles, execute command `extended test - cypress - record` or `extended test - cypress - record *`

`<AIO_version>` in commands is used to start test with multiple AIO instances - each instance in triggered by one jenkins build

-   To run with `master` and `stable`, execute command `extended test - matrix-test master,stable` or `extended test - matrix-test latest`
-   To run with specific version,ex: `2.3.0` and `2.3.1`, execute command `extended test - matrix-test 2.3.0,2.3.1`
-   In case `<AIO_version>` is empty, read versions from file `compTigerVersions.txt` of this repo

---

# PR Checklist

-   [ ] commit messages adhere to the [commit message guidelines](https://github.com/gooddata/gooddata-ui-sdk/blob/master/docs/contributing.md#what-should-the-commits-look-like)
-   [ ] review was done by a Code owner [if necessary](https://github.com/gooddata/gooddata-ui-sdk/blob/master/docs/contributing.md#how-do-i-tell-if-my-pull-request-needs-approval-by-a-code-owner) (if you think it is not necessary, explain the reasoning in the description or in a comment)
-   [ ] `check` passes
-   [ ] `extended test - backstop` passes
-   [ ] `extended test - tiger-cypress - record` to record new mapping files (Tiger BE)
-   [ ] `extended test - cypress - record` to record new mapping files (Bear BE)
-   [ ] `extended test - tiger-cypress - isolated` passes
-   [ ] `extended test - cypress - isolated` passes
-   [ ] `extended test - tiger-cypress - integrated` passes
-   [ ] `rush change` [was run if applicable](https://github.com/gooddata/gooddata-ui-sdk/blob/master/docs/contributing.md#how-do-i-describe-my-changes-for-the-changelog)

### CODE REVIEW

**Comment by NestorEncinas** [comment]
extended test - backstop

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/15172/ : SUCCESS in 4s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/16713/ : SUCCESS in 12m 21s

**Comment by yenkins** [comment]
Link to Storybook test result: https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-backstop-tests/2418/artifact/libs/sdk-ui-tests/backstop/output/html-report/index.html

**Comment by gdgate** [comment]
Build succeeded (check-extended pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/15173/ : SUCCESS in 4s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-pull-request-dispatcher-pipeline/1236/ : SUCCESS in 28m 39s

**Comment by NestorEncinas** [comment]
extended test - cypress - isolated

**Comment by gdgate** [comment]
Build succeeded (check-extended pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/15174/ : SUCCESS in 4s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-pull-request-dispatcher-pipeline/1237/ : SUCCESS in 11m 56s

File: libs/sdk-ui-ext/src/internal/components/pluggableVisualizations/pivotTable/widthItemsHelpers.ts [hunk 1]
```diff
@@ -74,16 +76,31 @@ const matchesWidthItemFilters = (
     return true;
 };
 
-const containsMeasureLocator = (widthItem: IMeasureColumnWidthItem): boolean =>
-    widthItem.measureColumnWidthItem.locators.some((locator) => isMeasureLocator(locator));
+function getMeasureOrTransposedWidthItem(
```
Comment (ivan-nejezchleb) [inline libs/sdk-ui-ext/src/internal/components/pluggableVisualizations/pivotTable/widthItemsHelpers.ts:79]: rename to getMeasureOrTransposedMeasureWidthItem

File: libs/sdk-ui-ext/src/internal/components/pluggableVisualizations/pivotTable/pivotTableAdditionalFactories.ts [hunk 2]
```diff
@@ -59,13 +62,48 @@ export function factoryNotationForMeasureColumnWidthItem(obj: IMeasureColumnWidt
...
     const { attributeIdentifier, element } = obj.attributeLocatorItem;
     // cannot use lodash compact, that would remove 0 values which we want to keep here
     const params = [`"${attributeIdentifier}"`, element && `"${element}"`].filter((item) => !isNil(item));
     return `newAttributeColumnLocator(${params.join(", ")})`;
 }
 
+export function factoryNotationForMeasureColumnLocator(obj: IMeasureColumnLocator): string {
+    const { measureIdentifier } = obj.measureLocatorItem;
+    // cannot use lodash compact, that would remove 0 values which we want to keep here
+    const params = [`"${measureIdentifier}"`].filter((item) => !isNil(item));
```
Comment (ivan-nejezchleb) [inline libs/sdk-ui-ext/src/internal/components/pluggableVisualizations/pivotTable/pivotTableAdditionalFactories.ts:103]: single item in array -> simplify code

**Review by ivan-nejezchleb — CHANGES_REQUESTED** [review]

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/15192/ : SUCCESS in 4s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/16725/ : FAILURE in 2m 48s

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/15193/ : SUCCESS in 4s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/16726/ : FAILURE in 2m 42s

File: libs/sdk-ui-ext/src/internal/components/pluggableVisualizations/pivotTable/pivotTableAdditionalFactories.ts [hunk 3]
```diff
@@ -59,13 +62,48 @@ export function factoryNotationForMeasureColumnWidthItem(obj: IMeasureColumnWidt
...
+): string {
+    const { locators, width } = obj.transposedMeasureColumnWidthItem;
+
+    const [measureLocators, attributeLocators] = partition<ColumnLocator, IMeasureColumnLocator>(
+        locators,
+        isMeasureColumnLocator,
+    );
+
+    const mergedLocators = [...measureLocators, ...attributeLocators];
+    const allowGrowToFit = isString(width) ? false : (width as IAbsoluteColumnWidth).allowGrowToFit;
```
[the remark this answers is not in the document]
Comment (NestorEncinas) [reply libs/sdk-ui-ext/src/internal/components/pluggableVisualizations/pivotTable/pivotTableAdditionalFactories.ts]: No longer needed.

File: libs/sdk-ui-ext/src/internal/components/pluggableVisualizations/pivotTable/pivotTableAdditionalFactories.ts (hunk 2 above)
Comment (NestorEncinas) [reply libs/sdk-ui-ext/src/internal/components/pluggableVisualizations/pivotTable/pivotTableAdditionalFactories.ts]: No longer needed.

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/15194/ : SUCCESS in 4s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/16727/ : FAILURE in 9m 42s

**Comment by gdgate** [comment]
Merge Failed.

Zuul merger could not merge this change into the base branch. This is most likely caused by merge conflicts. Please rebase the change and upload the rebased version. In case of further errors, contact the zuul administrator.

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/15196/ : SUCCESS in 4s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/16729/ : SUCCESS in 9m 44s

**Comment by NestorEncinas** [comment]
extended test - tiger-cypress - isolated

**Comment by gdgate** [comment]
Build succeeded (check-extended pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/15252/ : SUCCESS in 6s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-pull-request-dispatcher-pipeline/1252/ : SUCCESS in 8m 18s

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/15281/ : SUCCESS in 4s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/16811/ : FAILURE in 2m 38s

**Comment by gdgate** [comment]
Build in pipeline check aborted.

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/15283/ : SUCCESS in 4s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/16813/ : FAILURE in 9m 47s

**Comment by gdgate** [comment]
Build in pipeline check aborted.

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/15286/ : SUCCESS in 6s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/16816/ : SUCCESS in 11m 19s

**Comment by NestorEncinas** [comment]
extended test - cypress - isolated

**Comment by gdgate** [comment]
Build succeeded (check-extended pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/15287/ : SUCCESS in 4s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-pull-request-dispatcher-pipeline/1259/ : SUCCESS in 10m 57s

**Comment by NestorEncinas** [comment]
extended test - tiger-cypress - isolated

File: libs/sdk-ui-pivot/src/columnWidths.ts [hunk 4]
```diff
@@ -459,19 +478,49 @@ export function newWidthForAllColumnsForMeasure(
  * @remarks
  * See also {@link newAttributeColumnLocator} to learn more about the attribute column locators.
  *
- * @param measureOrId - Measure specified either by value or by localId reference
+ * @param measureOrId - Measures specified either by value or by localId reference
```
Comment (ivan-nejezchleb) [inline libs/sdk-ui-pivot/src/columnWidths.ts:481]: old function really works only with single measure

**Comment by gdgate** [comment]
Build in pipeline check aborted.

**Comment by yenkins** [comment]
### Sonar scan result


More detail, see in https://sonarqube-gate.intgdc.com/dashboard?id=gooddata-ui-sdk-gate-PR3989

To scan for vulnerabilities in dependencies and run unit tests (to get coverage report in sonar) please comment your PR with 'extended check sonar'.

**Comment by gdgate** [comment]
Build succeeded (gate pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-sonarqube-scan-zuul-gate/3103/ : SUCCESS in 10m 25s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/16825/ : SUCCESS in 9m 42s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-cypress-bear-isolated-zuul-pipeline/679/ : SUCCESS in 11m 22s
- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-merge-compliance-check/4261/ : SUCCESS in 10s
- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-fossa-scanning-zuul/983/ : FAILURE in 1m 12s (non-voting)
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-cypress-tiger-isolated-zuul-pipeline/619/ : SUCCESS in 11m 03s

**Pull request merged**